src/security/intel/cbnt/measurement.c

```diff
@@ -0,0 +1,543 @@
+	}
+
+	if (i == km->key_count) {
+		cbfs_unmap(km);
+		printk(BIOS_ERR, "CBnT: failed to find hash of BPM pubkey in KM\n");
+		return CB_ERR;
+	}
+
+	/* KM.BPKey.hashBuffer[BPM pubkey digest size] */
+	int err = obuf_write(ob, key_hash->hash.data, key_hash->hash.size);
+	if (err)
+		printk(BIOS_ERR, "CBnT: failed to write ACM signature\n");
+
+	cbfs_unmap(km);
+	return err ? CB_ERR : CB_SUCCESS;
+}
+
+static enum cb_err fill_pcr0_bpm_fields(struct obuf *ob, bool *auth_measure)
+{
+	size_t bpm_len;
+	struct bpm_header *bpm = map_cbfs_file(INTEL_CBNT_CBFS_BPM, &bpm_len);
+
+	struct ibuf bpm_sig_ib;
+	ibuf_init(&bpm_sig_ib, (uint8_t *)bpm + bpm->key_sig_offset,
+		  bpm_len - bpm->key_sig_offset);
+
+	/* BPM.Signature[BPM signature size] */
+	if (copy_signature(bpm_sig_ib, ob) != CB_SUCCESS) {
+		printk(BIOS_ERR, "CBnT: failed to copy BPM signature\n");
+		cbfs_unmap(bpm);
+		return CB_ERR;
+	}
+
+	const struct bpm_ibbs *ibbs = (const void *)bpm->se_element;
+	*auth_measure = (ibbs->flags & BPM_IBBS_FLAG_AUTH_MEASURE) != 0;
+
+	/* Auth data is measured only for TPM2. */
+	if (*auth_measure && tlcl_get_family() != TPM_2)
+		*auth_measure = false;
+
+	unsigned int i;
+	const struct hash_struct *ibb_hash = ibbs->digest_list.hashes;
+	for (i = 0; i < ibbs->digest_list.count; ++i) {
+		if (ibb_hash->alg == tpm2_alg_from_vb2_hash(tpm_log_alg()))
+			break;
+		ibb_hash = (const void *)&ibb_hash->data[ibb_hash->size];
+	}
+
+	if (i == ibbs->digest_list.count) {
+		printk(BIOS_ERR, "CBnT: failed to find matching IBB signature\n");
+		cbfs_unmap(bpm);
+		return CB_ERR;
+	}
+
+	/* IBB.Digest[IBB digest size] */
+	if (obuf_write(ob, ibb_hash->data, ibb_hash->size)) {
+		printk(BIOS_ERR, "CBnT: failed to write IBB signature\n");
+		cbfs_unmap(bpm);
+		return CB_ERR;
+	}
+
+	cbfs_unmap(bpm);
+	return CB_SUCCESS;
+}
+
+/*
+ * Pseudo-code of the data to be measured into PCR-7 for TigerLake and newer (older hardware
+ * isn't supported):
+ *
+ *     struct {
+ *         uint64_t ACM_POLICY_STATUS;
+ *         uint16_t ACM.Header.SVN;
+ *         uint8_t  ACM.KeyHash[32];
+ *         uint8_t  BP.KEY[32];
+ *         uint8_t  KM.BPKey.hashBuffer[BPM pubkey digest size];
+ *     } PCR7_DATA
+ *
+ * Applies only for TPM2.0.
+ *
+ * Returns true and initializes *hash on success.
+ */
+static bool make_pcr7_hash(struct obuf *ob, struct vb2_hash *hash)
+{
+	/*
+	 * ACM.KeyHash[32]
+	 *
+	 * BWG says to use TXT.PUBLIC.KEY, but TXT SDG (#315168-017.4) says that it
+	 * doesn't contain anything useful on modern platforms and suggests SGX-related
+	 * MSRs (0x20::0x23, to be specific, but SDM vol. 4 has these instead).
+	 */
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH0).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH1).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH2).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH3).raw);
+
+	/* BP.KEY[32] (rather its SHA256 hash) */
```

Comment:
Hash algorithm of this file is microachitecture dependent. TGL and onward already provision ME with SHA384 of the key, so the hash read from ME will be SHA384.

```suggestion
	/* BP.KEY[32] (rather its hash) */
```

Reply:
Updated: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R376